src/client-side-metrics/metrics-collector.ts

```diff
@@ -0,0 +1,329 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as fs from 'fs';
+import {IMetricsHandler} from './metrics-handler';
+import {
+  AttemptOnlyAttributes,
+  OnAttemptCompleteInfo,
+  OnOperationCompleteAttributes,
+  OperationOnlyAttributes,
+} from '../../common/client-side-metrics-attributes';
+
+/**
+ * An interface representing a Date-like object.  Provides a `getTime` method
+ * for retrieving the time value in milliseconds.  Used for abstracting time
+ * in tests.
+ */
+interface DateLike {
+  /**
+   * Returns the time value in milliseconds.
+   * @returns The time value in milliseconds.
+   */
+  getTime(): number;
+}
+
+/**
+ * Interface for a provider that returns DateLike objects. Used for mocking dates in tests.
+ */
+interface DateProvider {
+  /**
+   * Returns a DateLike object.
+   * @returns A DateLike object representing the current time or a fake time value.
+   */
+  getDate(): DateLike;
+}
+
+/**
+ * The default DateProvider implementation.  Returns the current date and time.
+ */
+class DefaultDateProvider {
+  /**
+   * Returns a new Date object representing the current time.
+   * @returns {Date} The current date and time.
+   */
+  getDate() {
+    return new Date();
+  }
+}
+
+/**
+ * An interface representing a tabular API surface, such as a Bigtable table.
+ */
+export interface ITabularApiSurface {
+  instance: {
+    id: string;
+  };
+  id: string;
+  bigtable: {
+    appProfileId?: string;
+  };
+}
+
+const packageJSON = fs.readFileSync('package.json');
+const version = JSON.parse(packageJSON.toString()).version;
+
+/**
+ * A class for tracing and recording client-side metrics related to Bigtable operations.
+ */
+export class MetricsCollector {
+  private operationStartTime: DateLike | null;
+  private attemptStartTime: DateLike | null;
+  private zone: string | undefined;
+  private cluster: string | undefined;
+  private tabularApiSurface: ITabularApiSurface;
+  private methodName: string;
+  private projectId?: string;
+  private attemptCount = 0;
```

Comment:
is a MetricsCollector created for each operation? 

If it's not global, I'd probably suggest a different name, like `OperationMetricsCollector`. But this also makes me wonder why we need the operation attributes classes, since it seems like the same data is here?

Reply:
There are no operation attributes classes. Those are interfaces. They are for assigning the parameters types. They don't store data.

Reply:
I just renamed it to `OperationMetricsCollector`.